### ISSUE TITLE
adapter-cloudflare: serve prerendered pages and static assets using `_routes.json`

### ISSUE THREAD

**Issue by Mexican-Man** [sveltejs/kit#7298]
### Describe the problem

From what I've seen and tested (please correct me), sites deployed via Cloudflare Pages still process prerendered pages (`export const prerender = true;`) and static assets through `_worker.js`. This means that even requests to static resources contribute to the host's "function calls" budget.

### Describe the proposed solution

[This recently updated Cloudflare doc](https://developers.cloudflare.com/pages/platform/functions/billing/#will-requests-to-assets-be-charged) indicates that `_routes.json` can be used to bypass `_worker.js`, which typically handles all requests. By restructuring the build output folder (to serve static assets as they are normally routed in-app), we could take advantage of this feature. This would make adapter-cloudflare infinitely more viable for high traffic sites.

### Alternatives considered

- Using adapter-static, but that only works if *all* of your site is prerendered.
- Stricter rate limiting, but that's not scalable, reliable, or customizable (4 rules max on Cloudflare).
- Maybe surprise outages are OK ¯\\\_(ツ)\_/¯

### Importance

i cannot use SvelteKit without it

### Additional Information

I understand that traditionally, the contents of the build output folder is not meant to be exposed, and that `_worker.js` is meant to handle *every* request, even for static assets. However, I think that falling back onto Cloudflare Page's built-in file routing would be well worth it if it meant saving *actual* money on hosting costs (or in my case, having my site stay live for the whole day).

**Comment by MedLeon** [comment]
This is a blocking issue for me as well.

I do not think that the `_redirect ` file is the way to go. The whole concept of a monolith _worker.js file is just for migration and should be abondend at some point. 
The "new" way would be to deploy a `[[path]].js` file. I would recommend the Cloudflare adapter for Astro. There you got two modes and one of them produces a  `[[path]].js` file.

**Comment by Mexican-Man** [comment]
> I would recommend the Cloudflare adapter for Astro. There you got two modes and one of them produces a `[[path]].js` file.

That's good to know, I'll keep that in mind.

> I do not think that the `_redirect ` file is the way to go. The whole concept of a monolith _worker.js file is just for migration and should be abondend at some point. The "new" way would be to deploy a `[[path]].js` file.

I think this would require a very large breaking overhaul to how SvelteKit compiles. However, the `_routes.json` approach could work without redirects, if SvelteKit compiled a static page to `/thing/index.html` rather than `/thing.html` (ie. the outputted file goes to where you would expect to access it via URL).

**Comment by benmccann** [comment]
I think static assets were addressed in https://github.com/sveltejs/kit/pull/6530. Maybe prerendered pages weren't though?

**Comment by Mexican-Man** [comment]
> I think static assets were addressed in #6530. Maybe prerendered pages weren't though?

[Here](https://github.com/sveltejs/kit/pull/6530/files#diff-78806028d48ad47894445421562b7796c53178948c017e13d4550def4193903eR90), it looks like static assets are *not* (unless favicon). Check out the comment there, I'm not really sure what I would do in that case. On the other hand, prerendered pages wouldn't work anyway, because they're not being outputted to the right folder/name.

In my fork, I worked around the assets (literally just comment the line), but working around the prerendered pages is janky.

In short, we could add static assets, but that would [disregard the comment made by the Pages crew](https://github.com/sveltejs/kit/pull/6530/files#diff-78806028d48ad47894445421562b7796c53178948c017e13d4550def4193903eR90). We could add prerendered pages, but in a way that's janky/messes with stuff.

**Comment by Mexican-Man** [comment]
I've redone my branch and modified how static pages are outputted. Interestingly, it seems like they were intentionally set up this way(?). [See my commits here](https://github.com/sveltejs/kit/compare/master...Mexican-Man:kit:prerender-routes-json). Couldn't say why, no one seems to be chiming in.

My fork seems to be working and passes all the tests, so I'm just going to use that for now. If no one has anything else to add, I'll submit a PR. As for static assets, at least those are cached by Cloudflare ¯\\\_(ツ)\_/¯

**Comment by benmccann** [comment]
maybe related to https://github.com/sveltejs/kit/issues/7640 which @jrf0110 and @Rich-Harris had begun to look at a bit?

**Comment by The-Noah** [comment]
Regardless of what is implemented, I think it would be extremely invaluable to be able to override - weather it's manually adding paths or removing paths that would be automatically added.

**Comment by Rich-Harris** [comment]
> The "new" way would be to deploy a `[[path]].js` file

To be clear, this is an authoring convenience only. Under the hood, your `[[path]].js` files get bundled into a `_worker.js` (unless the behaviour of Cloudflare Pages has changed since we last investigated this). SvelteKit is quite capable of emitting individual functions per route (it can do so for Vercel and Netlify with the `split: true` option), but there's no advantage to doing it on Cloudflare Pages.

> I've redone my branch and modified how static pages are outputted. Interestingly, it seems like they were intentionally set up this way(?). [See my commits here](https://github.com/sveltejs/kit/compare/master...Mexican-Man:kit:prerender-routes-json). Couldn't say why, no one seems to be chiming in.

See this conversation: https://github.com/sveltejs/kit/pull/6530#discussion_r962073965. The tl;dr is that someone might add their own `_middleware.ts` file, and @jrf0110 didn't want to prevent people from being able to run code in front of their assets. Personally I'd advocate for all assets being excluded, since that's how SvelteKit works on other platforms, but there's a maximum number of rules in `_routes.json` and it would be easy to exceed the limit if all static assets were included.

**Comment by MedLeon** [comment]
> unless the behaviour of Cloudflare Pages has changed since we last investigated this

It didn´t. I was wrong. You were right. Sorry.
It really seems like _routes.json is the only way to bypass functions.

**Comment by LoaiDev** [comment]
what is the reason for creating a single _worker.js instead of a function for each endpoint

**Comment by Mexican-Man** [comment]
> or will github count a request to a file which didn't invoke a function (static) towards the usage anyways?

Cloudflare charges per "function" aka per Worker invocation. Requests to static assets on the site are not billable.

> what is the reason for creating a single _worker.js instead of a function for each endpoint?

Consistent behaviour across platforms. Cloudflare Pages might serve assets, match paths, match URLs, etc. differently from the way SvelteKit wants to. That being said, it's easier just to let SvelteKit handle *all* of it, even if it causes extra Worker invocations (billable).

For example, Cloudflare Pages lets you have a static `404.html` page. If I wanted to do that with SvelteKit, I'd have to implement an `+error.svelte` file. The first approach is nice because it doesn't invoke a Worker but cannot be dynamic. The second approach is the opposite.

> won't that mean cloudflare will serve the static assets on its own?

They get served the same way, and still get cached (luckily). The difference, as Rich mentioned above, is that every call to a static asset now invokes a Worker.

**Comment by benmccann** [comment]
I believe this was just implemented in https://github.com/sveltejs/kit/pull/8422
